tools/base/fpgainfo/errors.c

```diff
@@ -0,0 +1,480 @@
+{
+	int i;
+	int j;
+	fpga_result res = FPGA_OK;
+	fpga_objtype objtype;
+	const char *const *error_string = NULL;
+	int size = 0;
+
+	if ((NULL == errinfos) || (0 == num_errors)) {
+		return;
+	}
+
+	if (errors_config.clear) {
+		for (i = 0; i < errors_config.force_count; i++) {
+			fpgaClearAllErrors(token);
+		}
+	}
+
+	res = fpgaPropertiesGetObjectType(props, &objtype);
+	fpgainfo_print_err("reading objtype from properties", res);
+
+	if (((VERB_ALL == errors_config.which)
+	     || (VERB_FME == errors_config.which))
+	    && (FPGA_DEVICE == objtype)) {
+		fpgainfo_print_common("//****** FME ERRORS ******//", props);
+
+		for (i = 0; i < (int)num_errors; i++) {
+			uint64_t error_value = 0;
+
+			res = fpgaReadError(token, i, &error_value);
+			fpgainfo_print_err("reading error for FME", res);
+
+			printf("%-29s : 0x%" PRIX64 "\n", errinfos[i].name,
+			       error_value);
+
+			if (!strcmp(errinfos[i].name, "Errors")) {
+				size = FME_ERROR_COUNT;
+				error_string = FME_ERROR;
+			} else if (!strcmp(errinfos[i].name, "Next Error")) {
+				size = 0;
+				error_string = NULL;
+			} else if (!strcmp(errinfos[i].name, "First Error")) {
+				size = 0;
+				error_string = NULL;
+			} else if (!strcmp(errinfos[i].name, "PCIe0 Errors")) {
+				size = PCIE0_ERROR_COUNT;
+				error_string = PCIE0_ERROR;
+			} else if (!strcmp(errinfos[i].name, "Inject Error")) {
+				size = INJECT_ERROR_COUNT;
+				error_string = INJECT_ERROR;
+			} else if (!strcmp(errinfos[i].name,
+					   "Catfatal Errors")) {
+				size = CATFATAL_ERROR_COUNT;
+				error_string = CATFATAL_ERROR;
+			} else if (!strcmp(errinfos[i].name,
+					   "Nonfatal Errors")) {
+				size = NONFATAL_ERROR_COUNT;
+				error_string = NONFATAL_ERROR;
+			} else if (!strcmp(errinfos[i].name, "PCIe1 Errors")) {
+				size = PCIE1_ERROR_COUNT;
+				error_string = PCIE1_ERROR;
+			}
+
+			for (j = 0; (j < size) && (NULL != error_string); j++) {
+				if (FPGA_BIT_IS_SET(error_value, j)) {
+					printf("\t %s \n", error_string[j]);
+				}
+			}
+		}
+	} else if (((VERB_ALL == errors_config.which)
+		    || (VERB_PORT == errors_config.which))
+		   && (FPGA_ACCELERATOR == objtype)) {
+		fpgainfo_print_common("//****** PORT ERRORS ******//", props);
+
+		for (i = 0; i < (int)num_errors; i++) {
+			uint64_t error_value = 0;
+			res = fpgaReadError(token, i, &error_value);
+			fpgainfo_print_err("reading error for PORT", res);
+
+			printf("%-29s : 0x%" PRIX64 "\n", errinfos[i].name,
+			       error_value);
+
+			if (!strcmp(errinfos[i].name, "Errors")) {
+				size = PORT_ERROR_COUNT;
+				error_string = PORT_ERROR;
+			} else if (!strcmp(errinfos[i].name,
+					   "First Malformed Req")) {
+				size = 0;
+				error_string = NULL;
+			} else if (!strcmp(errinfos[i].name, "First Error")) {
+				size = 0;
+				error_string = NULL;
+			}
+
+			for (j = 0; (j < size) && (NULL != error_string); j++) {
+				if (FPGA_BIT_IS_SET(error_value, j)) {
+					printf("\t %s \n", error_string[j]);
+				}
+			}
```

Comment:
Same here, let's think about how this can be refactored into a common function (394-418)